File: 3-rucksack-reorganization/python/main.py
from functools import reduce
from pathlib import Path
import string
from typing import Generator

TESTDATA = Path(__file__).parent.parent / "test_data.txt"
DATA = Path(__file__).parent.parent / "data.txt"


def read_data(file) -> Generator[str, None, None]:
    with open(file, "r") as f:
        while line := f.readline():
            yield line


def read_data_part_2(file, n: int = 3) -> Generator[str, None, None]:
    with open(file, "r") as f:
        lines = []
        while line := f.readline():
            lines.append(line)
            if len(lines) == n:
                yield "".join(lines)
                lines = []


def priority(letter: str) -> int:
    return 1 + string.ascii_letters.index(letter)


def process_line(line: str) -> int:
    letter = get_letter_from_line(line)
    return priority(letter)


def get_letter_from_line(line: str) -> str:
    line = line.strip()
    n_len = len(line)
    assert n_len != 0, "Line is empty"
    assert (
        n_len % 2 == 0
    ), f"Line does not have an even number of letters {n_len}: {line}"
    # find the item that appears in both compartments
    # get the set of letters in half the line
    s1 = set(line[: n_len // 2])
    # get the set of letters in the other half of the line
    s2 = set(line[n_len // 2 :])
    common = s1 & s2  # gets intersection between two sets
    assert len(common) == 1, f"One letter {common} not found in rucksack: {line}"
    return common.pop()


def reduce_lines_to_common_letters(lines: Generator[str, None, None]) -> set[str]:
    # convert line to list of chars and then to a set
    tot = set(list(next(lines)))
    return reduce(lambda tot, line: tot & set(list(line)), lines, tot)


def get_badge_from_lines(lines_txt: str) -> str:
    """
    Find the letter that appears in all lines
    """
    lines = (line for line in lines_txt.split("\n") if line)
    s = reduce_lines_to_common_letters(lines)
    assert len(s) == 1, breakpoint()  # "More than one letter found in group of elves"
    return s.pop()


def process_lines(lines: str) -> int:
    badge = get_badge_from_lines(lines)
    assert len(badge) == 1, "More than one badge found for the group"
    return priority(badge)


def part1(file) -> int:
    """
    Find the item type that appears in both compartments of each
    rucksack. What is the sum of the priorities of those item types?
    """
    return reduce(lambda tot, line: tot + process_line(line), read_data(file), 0)


def part2(file) -> int:
    """
    Find the item type that corresponds to the badges of each three-Elf group.
    What is the sum of the priorities of those item types?
    """
    # read in three lines at a time
    return reduce(
        lambda tot, lines: tot + process_lines(lines), read_data_part_2(file), 0
    )


def main():
    res = part1(TESTDATA)
    assert res == 157
    print("Part 1 test:", res)
    res = part1(DATA)
    print("Part 1:", res)

    # Part 2
    res = part2(TESTDATA)
    assert res == 70
    print("Part 2 test:", res)
    res = part2(DATA)
    print("Part 2:", res)


if __name__ == "__main__":
    main()

File: 3-rucksack-reorganization/rust/src/main.rs
use std::{
    fs,
    io::{BufRead, BufReader, Lines},
};

fn main() {
    let file = "../data.txt".to_string();
    let testfile = "../test_data.txt".to_string();
    let res = part1(&testfile);
    println!("Part 1 test: {res}");
    let res = part1(&file);
    println!("Part 2: {res}");
    let res = part2(&testfile);
    println!("Part 2 test: {res}");
    let res = part2(&file);
    println!("Part 2: {res}");
}

fn create_buffer(file: &str) -> BufReader<fs::File> {
    let file = fs::File::open(file).expect("Unable to open file");
    BufReader::new(file)
}

fn read_data(file: &str) -> Lines<BufReader<fs::File>> {
    let reader = create_buffer(file);
    reader.lines()
}

fn part1(file: &str) -> i32 {
    let mut tot = 0;
    let data = read_data(file);
    for line in data {
        match line {
            Ok(val) => {
                tot = tot + process_line_p1(val);
            },
            Err(e) => panic!("Error: {}", e),
        }
    }
    tot
}
fn part2(file: &str) -> i32 {
    let mut buf = create_buffer(&file);
    let mut tot = 0;
    // loop until buf is empty
    while buf.fill_buf().unwrap().len() > 0 {
        let mut line1 = String::new();
        let mut line2 = String::new();
        let mut line3 = String::new();
        // read from buffer three times filling each line
        buf.read_line(&mut line1).expect("Unable to read line");
        buf.read_line(&mut line2).expect("Unable to read line");
        buf.read_line(&mut line3).expect("Unable to read line");
        // process each line1, line2, line3
        tot = tot + process_lines_p2(line1, line2, line3);
    }
    tot
}

fn priority(c: char) -> i32 {
    // lowercase items a-z have priorities 1-26
    // uppercase items A-Z have priorities 27-52
    // return the priority of c
    let mut p = 0;
    if c.is_lowercase() {
        p = c as i32 - 96;
    } else if c.is_uppercase() {
        p = c as i32 - 64 + 26;
    }
    p
}

fn process_line_p1(line: String) -> i32 {
    // get the length of the line
    let len = line.len();
    // split the line in two equal halves
    let (first, second) = line.split_at(len / 2);
    // find the common character between first and second
    let common: char = find_first_common_char(first, second).unwrap();
    priority(common)
}

fn process_lines_p2(line1: String, line2: String, line3: String) -> i32 {
    // find the common characters between line1, line2, and line3
    let common = match find_common_chars(line1, line2, line3) {
        Ok(val) => val,
        Err(e) => panic!("Error: {}", e),
    };
    priority(common)
}

fn find_first_common_char<'a>(first: &'a str, second: &str) -> Result<char, &'a str> {
    for c in first.chars() {
        if second.contains(c) {
            return Ok(c);
        }
    }
    Err("No common character found")
}

fn find_common_chars(line1: String, line2: String, line3: String) -> Result<char, &'static str> {
    // find the common characters between line1, line2, and line3
    for c in line1.chars() {
        if line2.contains(c) && line3.contains(c) {
            let common = c;
            return Ok(common)
        }
    }
    Err("Cannot find a common character between the three lines")
}
